apps/server/src/modules/tool/launch-tool/service/strategy/abstract-launch.strategy.ts

```diff
@@ -0,0 +1,203 @@
+			properties: [],
+			openNewTab: externalToolDO.openNewTab,
+		});
+
+		return launchData;
+	}
+
+	private buildToolLaunchDataFromTools(data: IToolLaunchParams): PropertyDataDO[] {
+		const propertyData: PropertyDataDO[] = [];
+		const { externalToolDO, schoolExternalToolDO, contextExternalToolDO } = data;
+		const externalParameters = externalToolDO.parameters || [];
+
+		const scopes: { scope: CustomParameterScope; params: CustomParameterEntryDO[] }[] = [
+			{ scope: CustomParameterScope.GLOBAL, params: externalToolDO.parameters || [] },
+			{ scope: CustomParameterScope.SCHOOL, params: schoolExternalToolDO.parameters || [] },
+			{ scope: CustomParameterScope.CONTEXT, params: contextExternalToolDO.parameters || [] },
+		];
+
+		this.addParameters(propertyData, externalParameters, scopes, schoolExternalToolDO, contextExternalToolDO);
+
+		return propertyData;
+	}
+
+	private addParameters(
+		propertyData: PropertyDataDO[],
+		customParameterDOs: CustomParameterDO[],
+		scopes: { scope: CustomParameterScope; params: CustomParameterEntryDO[] }[],
+		schoolExternalToolDO: SchoolExternalToolDO,
+		contextExternalToolDO: ContextExternalToolDO
+	): void {
+		for (const { scope, params } of scopes) {
+			const parameterNames: string[] = params.map((parameter: CustomParameterEntryDO) => parameter.name);
+
+			const parametersToInclude: CustomParameterDO[] = customParameterDOs.filter(
+				(parameter: CustomParameterDO) => parameter.scope === scope && parameterNames.includes(parameter.name)
+			);
+
+			this.handleParametersToInclude(
+				parametersToInclude,
+				params,
+				propertyData,
+				schoolExternalToolDO,
+				contextExternalToolDO
+			);
+		}
+	}
+
+	private handleParametersToInclude(
+		parametersToInclude: CustomParameterDO[],
+		params: CustomParameterEntryDO[],
+		propertyData: PropertyDataDO[],
+		schoolExternalToolDO: SchoolExternalToolDO,
+		contextExternalToolDO: ContextExternalToolDO
+	): void {
+		for (const parameter of parametersToInclude) {
+			const matchingParameter: CustomParameterEntryDO | undefined = params.find(
+				(param: CustomParameterEntryDO) => param.name === parameter.name
+			);
+
+			if (matchingParameter) {
+				const value = this.getParameterValue(parameter, matchingParameter, schoolExternalToolDO, contextExternalToolDO);
+
+				if (value !== undefined) {
+					this.addProperty(propertyData, parameter.name, value, parameter.location);
+				}
+			}
+		}
+	}
+
+	private getParameterValue(
+		customParameter: CustomParameterDO,
+		matchingParameterEntry: CustomParameterEntryDO,
+		schoolExternalToolDO: SchoolExternalToolDO,
+		contextExternalToolDO: ContextExternalToolDO
+	): string | undefined {
+		if (customParameter.type === CustomParameterType.AUTO_SCHOOLID) {
+			return schoolExternalToolDO.schoolId;
+		}
+
+		if (
+			customParameter.type === CustomParameterType.AUTO_COURSEID &&
+			contextExternalToolDO.contextType === ToolContextType.COURSE
+		) {
+			return contextExternalToolDO.contextId;
+		}
```

Comment:
OK for now, but we need to find a more generic solution once we add more AUTO params. Maybe comment this for further reference

Reply:
Yes, should we do maybe in a seperate service in the future if it will more complicated